ld.script

```diff
@@ -0,0 +1,233 @@
+  .got.plt        : { *(.got.plt)  *(.igot.plt) }
+  .data           :
+  {
+    *(.data .data.* .gnu.linkonce.d.*)
+    SORT(CONSTRUCTORS)
+  }
+  .data1          : { *(.data1) }
+  _edata = .; PROVIDE (edata = .);
+  . = .;
+  __bss_start = .;
+  .bss            :
+  {
+   *(.dynbss)
+   *(.bss .bss.* .gnu.linkonce.b.*)
+   *(COMMON)
+   /* Align here to ensure that the .bss section occupies space up to
+      _end.  Align after .bss to ensure correct alignment even if the
+      .bss section disappears because there are no input sections.
+      FIXME: Why do we need it? When there is no .bss section, we don't
+      pad the .data section.  */
+   . = ALIGN(. != 0 ? 64 / 8 : 1);
+  }
+  .lbss   :
+  {
+    *(.dynlbss)
+    *(.lbss .lbss.* .gnu.linkonce.lb.*)
+    *(LARGE_COMMON)
+  }
+  . = ALIGN(64 / 8);
+  . = SEGMENT_START("ldata-segment", .);
+  .lrodata   ALIGN(CONSTANT (MAXPAGESIZE)) + (. & (CONSTANT (MAXPAGESIZE) - 1)) :
+  {
+    *(.lrodata .lrodata.* .gnu.linkonce.lr.*)
+  }
+  .ldata   ALIGN(CONSTANT (MAXPAGESIZE)) + (. & (CONSTANT (MAXPAGESIZE) - 1)) :
+  {
+    *(.ldata .ldata.* .gnu.linkonce.l.*)
+    . = ALIGN(. != 0 ? 64 / 8 : 1);
+  }
+  . = ALIGN(64 / 8);
+  _end = .; PROVIDE (end = .);
+  . = DATA_SEGMENT_END (.);
+  /* Stabs debugging sections.  */
+  .stab          0 : { *(.stab) }
+  .stabstr       0 : { *(.stabstr) }
+  .stab.excl     0 : { *(.stab.excl) }
+  .stab.exclstr  0 : { *(.stab.exclstr) }
+  .stab.index    0 : { *(.stab.index) }
+  .stab.indexstr 0 : { *(.stab.indexstr) }
+  .comment       0 : { *(.comment) }
+  /* DWARF debug sections.
+     Symbols in the DWARF debugging sections are relative to the beginning
+     of the section so we begin them at 0.  */
+  /* DWARF 1 */
+  .debug          0 : { *(.debug) }
+  .line           0 : { *(.line) }
+  /* GNU DWARF 1 extensions */
+  .debug_srcinfo  0 : { *(.debug_srcinfo) }
+  .debug_sfnames  0 : { *(.debug_sfnames) }
+  /* DWARF 1.1 and DWARF 2 */
+  .debug_aranges  0 : { *(.debug_aranges) }
+  .debug_pubnames 0 : { *(.debug_pubnames) }
+  /* DWARF 2 */
+  .debug_info     0 : { *(.debug_info .gnu.linkonce.wi.*) }
+  .debug_abbrev   0 : { *(.debug_abbrev) }
+  .debug_line     0 : { *(.debug_line .debug_line.* .debug_line_end ) }
+  .debug_frame    0 : { *(.debug_frame) }
+  .debug_str      0 : { *(.debug_str) }
+  .debug_loc      0 : { *(.debug_loc) }
+  .debug_macinfo  0 : { *(.debug_macinfo) }
+  /* SGI/MIPS DWARF 2 extensions */
+  .debug_weaknames 0 : { *(.debug_weaknames) }
+  .debug_funcnames 0 : { *(.debug_funcnames) }
+  .debug_typenames 0 : { *(.debug_typenames) }
+  .debug_varnames  0 : { *(.debug_varnames) }
+  /* DWARF 3 */
+  .debug_pubtypes 0 : { *(.debug_pubtypes) }
+  .debug_ranges   0 : { *(.debug_ranges) }
+  /* DWARF Extension.  */
+  .debug_macro    0 : { *(.debug_macro) }
+  .gnu.attributes 0 : { KEEP (*(.gnu.attributes)) }
+  /DISCARD/ : { *(.note.GNU-stack) *(.gnu_debuglink) *(.gnu.lto_*) }
+}
```

Comment:
This file can probably be reduced to ~20 lines.  It's pretty well hidden in all the definitions but I infer you really only care about the `nodetext` symbols.

Reply:
I dont see how it can be reduced to 20 lines. Much of it is what is used by the default linker script. Yes the only reason for me to modify it was to find the start of the actual text. I had trouble with moving the plt regions and wanted to avoid several of those regions.

Reply:
https://sourceware.org/binutils/docs/ld/Implicit-Linker-Scripts.html#Implicit-Linker-Scripts - instead of replacing the default linker script, you use an implicit linker script that works in addition to the default linker script.

Replacing the default linker script isn't an option, IMO, because that will be a never-ending maintenance hassle for us.

Reply:
The implicit linker scripts wont really help as I cant do a symbol assignment for something that is already specified in the linker script. I will investigate if there is some other approach to get the address of the .text segment.

Reply:
I am able eliminate the ld.script by using the LD option -Wl,-Ttext=<address to place text segment>. This places the text segment at a specific address which I can then use as the start for the 2M mapping. This will hard code the text starting address.

The other way is to read the elf file and find the .text segment and determine the address.

Reply:
A fixed address is bad for security reasons and reading it from the ELF header won't work when ASLR is enabled.

I guess you could compute the ASLR slide at run-time but that gets complicated fast when code and data is spread across several segments.

Another issue with that approach: two segments with different protection might be located within < 2 MB from each other.

Are you sure implicit linker scripts don't work?  I remember using that to put symbols at fixed locations.

What about retrieving the path to the default linker script with `ld --verbose` and passing `-T /path/to/default/script.ld -T custom.ld`?

Reply:
Thanks. Let me look into the linker script some more. One option is to generate the linker script using the `ld --verbose` and to then add my symbol to it.

Reply:
I was able to figure this out. GNU ld has the INSERT option which makes the externally supported script not override the default script but insert the code into the default. So I will delete the ld.script and insert a new ld.implicit.script

  PROVIDE (_nodetext = .);
  PROVIDE (nodetext = .);
  INSERT BEFORE .text;